### PR TITLE
Implement transform_constant_{and,or}_write_node

### PR DESCRIPTION
**Pull request by herwinw**
This resolves compilation errors with code like this:
```ruby
Foo ||= 1
Foo &&= 2
```

### CODE REVIEW

**Comment by herwinw** [comment]
Rebased and marked as ready for review. This change does not cover the path constants (like `A::B ||= 1`). When trying those, it's probably easier to start with the `transform_constant_path_and_write_node` operation, since the specs of the `or` equivalent really hit the bug in  #1734

**Comment by seven1m** [comment]
Dang I gotta look at that mac build

File: lib/natalie/compiler/pass1.rb [hunk 1]
```diff
@@ -844,6 +844,44 @@ def transform_class_variable_write_node(node, used:)
...
+        instructions << PopInstruction.new unless used
+        instructions
+      end
+
+      def transform_constant_or_write_node(node, used:)
+        instructions = [
+          IsDefinedInstruction.new(type: 'constant'),
+          PushSelfInstruction.new,
+          ConstFindInstruction.new(node.name, strict: false),
+          EndInstruction.new(:is_defined),
```
Comment (seven1m) [inline lib/natalie/compiler/pass1.rb:870]: I wonder why this one needs the IsDefinedInstruction but the other transform does not 🤔

Comment (herwinw) [reply lib/natalie/compiler/pass1.rb:870]: https://kddnewton.com/2023/12/06/advent-of-prism-part-6#constants has some explanation for that

**Pull request merged**